Fix part of #10474: Cover exploration-feature service with strict typescript checks

rodemore opened the pull request:
## Overview

1. This PR fixes or fixes part of #10474.
2. This PR does the following: cover ExplorationFeatureService with strict typescript checks

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers
- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

rodemore committed:
cover exploration-features.service.ts with stict checks

rodemore committed:
corrections pre-push

rodemore commented:
@nishantwrp  PTAL

kevintab95 reviewed (approved):
LGTM for codeowner file `core/templates/services/exploration-features-backend-api.service.ts`. Thanks @rodemore!

vojtechjelinek commented on tsconfig-strict.json:
```diff
@@ -111,6 +111,7 @@
     "core/templates/services/debouncer.service.spec.ts",
     "core/templates/services/editability.service.ts",
     "core/templates/services/editability.service.spec.ts",
+    "core/templates/services/exploration-features.service.ts",
```
Can you please type the spec file for exploration features too?

rodemore replied:
Done!

rodemore committed:
updating branch with the latest changes in develop

DubeySandeep commented on core/templates/services/utils.service.ts:
```diff
@@ -19,29 +19,27 @@
-   * @return {boolean} - true if object is empty, false otherwise.
-   */
+  * @param {Object} obj - the object to be checked.
+  * @return {boolean} - true if object is empty, false otherwise.
```
Undo these changes*

rodemore replied:
Done!

rodemore replied:
Done!

vojtechjelinek replied:
Not done.

rodemore replied:
In the commit "_removing unwanted changes from core/templates/services/utils.service.ts_" I restored the spaces so that they no longer differ from the develop branch.

vojtechjelinek replied:
No, you didn't this file still differs from develop.

rodemore replied:
Yeah. Sorry, I didn't notice the indentation problem. But now it's Done!

DubeySandeep commented on core/templates/services/utils.service.ts:
```diff
@@ -50,26 +48,24 @@ export class UtilsService {
-   * @param {Object} input - the object to be checked.
-   * @return {boolean} - true if input is string, false otherwise.
-   */
+  * @param {Object} input - the object to be checked.
```
ANy expected reason for these changes?

rodemore replied:
Unwanted changes was removing!

vojtechjelinek replied:
Not done.

rodemore replied:
In the commit "_removing unwanted changes from core/templates/services/utils.service.ts_" I restored the spaces so that they no longer differ from the develop branch.

DubeySandeep commented on core/templates/services/utils.service.ts:
```diff
@@ -109,7 +109,6 @@ export class UtilsService {
       default: return value instanceof Error;
     }
   }
-
```
Please undo unwanted changes!

rodemore replied:
Done!

rodemore replied:
Done!

DubeySandeep commented on core/templates/services/utils.service.ts:
```diff
@@ -120,7 +119,6 @@ export class UtilsService {
     }
   }
 }
-
```
Ditto*

rodemore replied:
Done!

DubeySandeep commented on core/templates/services/utils.service.ts:
```diff
@@ -88,19 +84,23 @@ export class UtilsService {
     for (var i = 0; i < aProps.length; i++) {
       var propName = aProps[i];
-      if (!this.isEquivalent(a[propName], b[propName])) {
+      const getKeyValue = (key: string) =>
+        (obj: Record<string, Object>) => obj[key];
+      if (!this.isEquivalent(
+        getKeyValue(propName)(<Record<string, object>>a),
+        getKeyValue(propName)(<Record<string, object>>b))
+      ) {
```
Can you please explain why we had to do it this way?

rodemore replied:
The problem I was getting in that case was this: 
_"Element implicitly has an 'any' type because expression of type 'string' can't be used to index type"._

I found the solution to this problem at https://github.com/microsoft/TypeScript/issues/35859 , but in order not to affect the structure of the Object interface I additionally parse the object to a Record<>

vojtechjelinek replied:
Could you add a comment in the code to explain this ((you can include the link to the issue).

rodemore replied:
Done!

DubeySandeep commented on core/templates/services/utils.service.ts:
```diff
@@ -19,29 +19,27 @@
 
 import { Injectable } from '@angular/core';
 import { downgradeInjectable } from '@angular/upgrade/static';
-
```
PLease remove unwanted changes. (Try to add minimal changes in the PR)

rodemore replied:
Done!

DubeySandeep commented on core/templates/services/contextual/url.service.ts:
```diff
@@ -149,15 +149,15 @@ export class UrlService {
   }
 
-  getClassroomUrlFragmentFromLearnerUrl(): string {
+  getClassroomUrlFragmentFromLearnerUrl(): string | undefined {
```
why the `get` method is returning `undefined`?

rodemore replied:
Same case as above.

rodemore replied:
In this function there is a case where it throws an error when the URL is invalid. In that case the typescript checks give me the warning that the function can return undefined.

srijanreddy98 replied:
Shouldn't it be void instead of undefined?

rodemore replied:
Both void and undefined work in this case. But maybe using void makes it more readable. Thanks!

rodemore replied:
I used _undefined_ because the return of void functions could be considered _undefined_. I had not tried _void_. But now changing it and doing the tests, I got this error in more than 10 files:

![image](https://user-images.githubusercontent.com/46513107/114818035-ab491400-9d80-11eb-9302-e7fbb50e8238.png)

Should I continue to use _undefined_, or should I change it to _void_ and fix the error I showed in all the files where this occurs?

rodemore replied:
PTAL @srijanreddy98

rodemore replied:
Here I try a function that doesn't have a "return", and the type I get is _undefined_

![image](https://user-images.githubusercontent.com/46513107/114819251-cd439600-9d82-11eb-9e62-72d1b118cd3e.png)

srijanreddy98 replied:
Thanks! Approved.

DubeySandeep commented on core/templates/services/contextual/url.service.ts:
```diff
@@ -149,15 +149,15 @@ export class UrlService {
   }
 
-  getSubtopicUrlFragmentFromLearnerUrl(): string {
+  getSubtopicUrlFragmentFromLearnerUrl(): string | undefined {
```
why the `get` method is returning `undefined`?

rodemore replied:
In this function there is a case where it throws an error when the URL is invalid. In that case the typescript checks give me the warning that the function can return undefined.

rodemore replied:
In this function there is a case where it throws an error when the URL is invalid. In that case the typescript checks give me the warning that the function can return undefined.

DubeySandeep commented on core/templates/services/contextual/url.service.ts:
```diff
@@ -107,7 +107,7 @@ export class UrlService {
    * @throws Will throw an error if the url is invalid.
    */
-  getTopicUrlFragmentFromLearnerUrl(): string {
+  getTopicUrlFragmentFromLearnerUrl(): string | undefined {
```
why the `get` method is returning `undefined`?

rodemore replied:
In this function there is a case where it throws an error when the URL is invalid. In that case the typescript checks give me the warning that the function can return undefined.

rodemore replied:
In this function there is a case where it throws an error when the URL is invalid. In that case the typescript checks give me the warning that the function can return undefined.

DubeySandeep commented on core/templates/services/contextual/url.service.ts:
```diff
@@ -61,7 +61,7 @@ export class UrlService {
   https://github.com/oppia/oppia/pull/7834#issuecomment-547896982 */
   getUrlParams(): UrlParamsType {
-    let params = {};
+    let params: Record<string, string> = {};
```
Please define an interface and have a name for the key and value. (Ditto here and elsewhere)

rodemore replied:
Done! I saw that there was an interface that I could reuse in that same file and I used it.

DubeySandeep reviewed (changes requested):
@rodemore Thanks for the PR, I took a pass and left a few comments, PTAL!

rodemore committed:
removing unwanted changes from core/templates/services/utils.service.ts

rodemore commented:
> Thanks for the PR, I took a pass and left a few comments, PTAL!

Thanks to you, this is my first PR and your comments will surely help me to improve in my next PRs.

rodemore committed:
address review comments

rodemore reviewed (commented):
@DubeySandeep  
I addressed the review comments

vojtechjelinek commented on core/templates/services/utils.service.ts:
```diff
@@ -66,10 +66,10 @@ export class UtilsService {
+  * @param {Object} b - the second object to be compared.
+  * @return {boolean} - true if a is equivalent to b, false otherwise.
+  */
   isEquivalent(a: Object, b: Object): boolean {
```
Can you please check the call sites of this function to see if we can maybe make the type definition a bit more tight?

rodemore replied:
I tried to find a way to set a more specific type.But that function is used in 8 files, in general it compares elements of type Object or other interfaces. For that reason I left them in type Object.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

rodemore committed:
add types for exploration-features.service.spec.ts

rodemore committed:
address review comments

rodemore committed:
removing unwanted changes

vojtechjelinek commented on core/templates/services/utils.service.ts:
```diff
@@ -86,9 +86,19 @@ export class UtilsService {
     if (aProps.length !== bProps.length) {
       return false;
     }
+    // In the indexing of an Object type by a string
+    // there is implicitly an 'any' type.
+    // Issue solved: https://github.com/microsoft/TypeScript/issues/35859
+    // Additionally a cast was added to the Record type in order not
+    // to modify the structure of the Object interface.
```
```suggestion
    // The indexing of an Object with a string implicitly returns 
    // 'any' type. This issue is solved according to
    // https://github.com/microsoft/TypeScript/issues/35859.
    // Additionally a cast was added to the Record type in order not
    // to modify the structure of the Object interface.
```

rodemore replied:
Done!

vojtechjelinek commented on core/templates/services/exploration-features-backend-api.service.ts:
```diff
@@ -32,7 +32,7 @@ interface ExplorationFeaturesBackendDict {
 export interface ExplorationFeatures {
   isExplorationWhitelisted: boolean;
-  alwaysAskLearnersForAnswerDetails: boolean;
+  alwaysAskLearnersForAnswerDetails?: boolean;
```
Why this needs to be optional?

rodemore replied:
![image](https://user-images.githubusercontent.com/46513107/113520312-c97c6c00-9557-11eb-979c-7093d873e51d.png)
This interface is used in the exploration-features.service.spec.ts, and the *alwaysAskLearnersForAnswerDetails* property is not there.

vojtechjelinek replied:
Hmm, okay, if it is just in tests then make sure to fill this out in the test files and not mark it as optional. This applies to any other places.

rodemore replied:
@vojtechjelinek Maybe like this?

![image](https://user-images.githubusercontent.com/46513107/113730290-60166d80-96bd-11eb-9d4c-3b4d129c1c73.png)

Adding comments to explain why that property is there

vojtechjelinek reviewed (changes requested):
Thanks! Just a few remaining comments.

rodemore committed:
update comments for exploration-features service

DubeySandeep reviewed (approved):
LGTM for the code owner file.

rodemore commented:
> Thanks! Just a few remaining comments.

It's done. PTAL @vojtechjelinek 
Thanks!

vojtechjelinek reviewed (changes requested):
Thanks! One remaining comment that I missed before (sorry for that).

rodemore committed:
address review comment

vojtechjelinek reviewed (approved):
LGTM!

rodemore commented:
@srijanreddy98  PTAL

rodemore committed:
update branch with the latest changes from develop

rodemore committed:
update branch

arpit1912 commented:
@srijanreddy98  PTAL!

rodemore committed:
Merge remote-tracking branch 'upstream/develop' into rodemore-typescript-checks

srijanreddy98 reviewed (changes requested):
Thanks @rodemore. Just one concern otherwise LGTM!

rodemore committed:
change 'undefined' to 'void' in url.service.ts

mridul-netizen committed:
Fix Part of #10700: Refractor Object Factories (#12458)

* migrated object factories

* updated

* removed unnecessary imports

* renamed files

* updated

* updated

* alphabetical order

* updated

* alphabetical order

* nits

Commit:
Add MachineTranslatedText storage model and domain object (#12515)

* workspace transfer

* formatting

* move MachineTranslatedText to translation_domain.py

* import utils convert to hash

* add storage model

* add translation domain object and tests

* add codeowners

* fix typo

* change code owner to @vojtechjelinek

* add translation_validators

* sort imports
fix spacing

* fix origin_text type
Change max hash length to be more "in range"
Fix/clarify docstrings and comments
Add TODOs for tidying language code verification logic related to #12341

* fix spacing

* add MachineTranslatedTextModelAuditOneOffJob to job registry

* fix typo

* add hashed_origin_text field to MachineTranslatedTextModel
(allows lookup by origin_text)

* address pr comments:
change MachineTranslatedTextTests name to be consistent
clarify/clean-up comments
remove unnecessary backslashes from regex
fix spacing

* fix typo

* add translation fetchers

* fix spacing

* fix misnamed setup
sort codeowners lines

* fix formating and docstring nits

* change _generate_id to be a static method

* add tests for get_translation_for_text
clarify docstrings

* consolidate terminology:
    replace "origin" with "source" in reference to the untranslated source text

* add more description regarding key purpose

* replace ":" with "." as a divider character to conform to codebase practices

Co-authored-by: Theo Lipeles <theolipeles@google.com>

rohitkatlaa committed:
Fix part of #11475: Migrate validate_model_domain_object_instances to beam jobs (#12317)

* migrating validate_model_domain_object_instances and validate_external_id_relationships to beam jobs

* migrating validate_model_domain_object_instances and validate_external_id_relationships to beam jobs

* adding tests

* adding tests

* finishing tests

* linting

* merge issues

* lint issues

* removing classes

* removing imports

* adding tests

* requested changes

* req changes

* requested change

ashutoshc8101 committed:
Fix part of #10474: Strict type checks for NeedsGuidingResponsesTaskModal and its spec file (#12491)

* done

* tests paddd

* test also done

* done

* sad

* m;

* done

* sad

auschoi committed:
changed test_accept_suggestion_invalid_suggestion_failure method to use public interface, added helper function to invalidate suggestion (#12511)

SAEb-ai committed:
Added .() (#12530)

brianrodri committed:
Fix part of #11462: Add one-off job to detect accounts with duplicate emails (#12534)

* add audit job

* fix lint

* fix typo

Commit:
Add Urdu to list of supported languages (#12535)

* Add Urdu to list of supported languages

* update SUPPORTED_CONTENT_LANGUAGES

srijanreddy98 committed:
Fix part of #9749: Migrate Custom OSK, drag-drop-positive, filepath, fraction, graph-editor, html-editor and image-with-regions editors (#12502)

* Algebriac migrated

* guppy

* boolean-editor migrated

* coord-two-dim

* Success!

* Check_fe

* Fix attribution

* Delete third-party

* Fix imports

* Fix lint errors

* Fix tests

* Changes

* Check_frontend

* Check Frontend

* Changes

* Half changes

* Half changes

* Introduce event bus

* Fix tests

* Update core/templates/oppia-events/event-bus.service.ts

Co-authored-by: Sandeep Dubey <dubeysandeep.in@gmail.com>

* Fix unit tests

* Lint checks

* Fraction Editor Migrated

* Fraction migrated

* Rename file

* Filepath editor-- half migrated

* Fix bugs

* Lint chceks

* Lint chceks

* Final Lint fixes 😑

* Final Fix

* Review comments

* Rename oppia-events to app events

* Fix unit test coverage

* Fix

* Migrated Image-regions

* fix lint errors

* Final lint checks

* Add new unit test

* Fix upstream merge

* Update extensions/interactions/GraphInput/directives/graph-viz.component.ts

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* Review

* Review

* Changes

* Change class to interface

Co-authored-by: Sandeep Dubey <dubeysandeep.in@gmail.com>
Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

rodemore committed:
Merge remote-tracking branc 'upstream/develop' into 'rodemore-typescript-checks'

rodemore committed:
Merge remote-tracking branch 'upstream/develop' into rodemore-typescript-checks

rodemore committed:
Merge branch 'rodemore-typescript-checks' of https://github.com/rodemore/oppia into rodemore-typescript-checks

arpit1912 commented:
@sajalasati it seems while updating the branch @rodemore  did some mistake and more reviewers are assigned to his PR,  should he created a new PR or get it reviewed here? Thanks!

rodemore commented:
@BenHenning @iamprayush @U8NWXD @srijanreddy98 PTAL

rodemore committed:
Merge remote-tracking branch 'upstream/develop' into rodemore-typescript-checks

BenHenning commented:
Unless I'm mistaken, it appears my codeowners review is no longer needed here. Please re-assign me if I'm mistaken or if you'd like my review regardless.

U8NWXD commented:
I'm deassigning myself since this PR doesn't appear to change any files I'm codeowner for. Please reassign me and let me know if I'm wrong or if there's another reason you want me to review!

DubeySandeep commented:
@srijanreddy98 Can you please review this PR?

rodemore committed:
Merge remote-tracking branch 'upstream/develop' into rodemore-typescript-checks

rodemore committed:
Merge remote-tracking branch 'upstream/develop' into rodemore-typescript-checks
